Split subset by group

Konboi opened the pull request:
Add `--split-by-groups` option to the `split-subset` command

e.g)

```sh
$ launchable record tests --group e2e ./e2e-test-result.xml
$ find ./tests/**/test_*.py  | launchable subset --split --target 40% file > subset_id.txt
$ launchable split-subset --subset-id $(cat subset-id.txt) --split-by-groups file
$ ls -l 
rest-e2e.txt
rest-nogroup.txt
subset-e2e.txt 
subset-groups.txt
subset-nogroup.txt
```

shibuiwilliam commented on launchable/commands/split_subset.py:
```diff
@@ -60,28 +84,33 @@ def __init__(self, dry_run: bool = False):
-                    err=True,
-                )
-                return
+            if not is_split_by_groups and bin_target is None:
```
[ask]
Is it ok to specify `is_split_by_groups` and `bin_target` simultaneously?

Konboi replied:
No, it isn't so I fixed it. Thank you!

shibuiwilliam commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +211,44 @@ def run(self):
+                        self.write_file("{}/rest-{}.txt".format(split_by_groups_output_dir, group_name), rests)
+
+                if len(group_names) > 0:
+                    open(
+                        "{}/{}".format(split_by_groups_output_dir, SPLIT_BY_GROUP_SUBSET_GROUPS_FILE_NAME), "w+",
+                        encoding="utf-8").write("\n".join(group_names))
```
[must]
Since Python won't close an open file automatically, it is recommended to open in `with` clause to avoid forgetting to close.

```python
with open(
      "{}/{}".format(split_by_groups_output_dir, SPLIT_BY_GROUP_SUBSET_GROUPS_FILE_NAME), "w+",
      encoding="utf-8") as f:
      f.write("\n".join(group_names))
```

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -60,28 +84,33 @@ def __init__(self, dry_run: bool = False):
-                return
+            if (not is_split_by_groups and bin_target is None) or (is_split_by_groups and bin_target):
+                raise click.BadOptionUsage("--bin or --split-by-groups", "Missing option '--bin' or '--split-by-groups'")
+
```
We can create `run_split_by_group` and immediately branch to that method at the beginning. Most of the code below are orthogonal to the binning option, and some of the behaviors do not make sense to this mode (e.g. the fallback code).

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +211,44 @@ def run(self):
+                    if is_output_exclusion_rules:
+                        subset, rests = rests, subset
+
+                    if len(subset) > 0:
+                        self.write_file("{}/subset-{}.txt".format(split_by_groups_output_dir, group_name), subset)
+                        if group_name != SPLIT_BY_GROUPS_NO_GROUP_NAME:
+                            group_names.append(group_name)
+
+                    if len(rests) > 0:
+                        self.write_file("{}/rest-{}.txt".format(split_by_groups_output_dir, group_name), rests)
```
This is not aligned with subset https://github.com/launchableinc/cli/blob/main/launchable/commands/subset.py#L184-L192

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +211,44 @@ def run(self):
+                        if group_name != SPLIT_BY_GROUPS_NO_GROUP_NAME:
+                            group_names.append(group_name)
+
+                    if len(rests) > 0:
```
Per spec, this should be produced only when --rest is used.

Konboi replied:
click v8 supports optional value https://click.palletsprojects.com/en/8.1.x/options/. 
However, we use click v7 when the python version is under 3.6 https://github.com/launchableinc/cli/blob/main/setup.cfg#L18-L20

Hum... how to implement the rest option with the split-by-groups option ...

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +211,44 @@ def run(self):
+                        self.write_file("{}/rest-{}.txt".format(split_by_groups_output_dir, group_name), rests)
+
+                if len(group_names) > 0:
+                    with open("{}/{}".format(split_by_groups_output_dir, SPLIT_BY_GROUP_SUBSET_GROUPS_FILE_NAME),
```
Per spec, rest-groups should be created.

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +211,44 @@ def run(self):
-                return
+            if is_split_by_groups:
+                group_names = []
+                for group in split_groups:
```
Per spec, nogroup file should be created.

draftcode replied:
Read this again. I read in a wrong way.

Konboi commented:
## Changes 

- Dividied methods to `split_by_bin` and `split_by_group_names`
- Added `--split-by-groups-with-rest` option
  - Because needs to pass PATH to `--rest` option, however split by groups won't use the path will be received from `--rest` option.
- Added `split_by_groups_output_handler` and `split_by_groups_exclusion_output_handler` properties to be able to edit output format.
  - Supported only gradle plugin

shibuiwilliam commented on launchable/commands/split_subset.py:
```diff
@@ -52,48 +82,74 @@ def split_subset(
+
     class SplitSubset(TestPathWriter):
         def __init__(self, dry_run: bool = False):
+            self.split_by_groups_output_handler = self._default_split_by_groups_output_handler
+            self.split_by_groups_exclusion_output_handler = self._default_split_by_groups_exclusion_output_handler
+            self.is_split_by_groups_with_rest = is_split_by_groups_with_rest
+            self.split_by_groups_output_dir = split_by_groups_output_dir
             super(SplitSubset, self).__init__(dry_run=dry_run)
```
[nits]
It's rather common to write `super().__init__()` first to override parent constructors. Even though it does not effect in this code, writing this phrase at last will make parent variables override child variables when the child class is meant override parant, which is nonsense.

shibuiwilliam replied:
ok

shibuiwilliam commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +250,76 @@ def run(self):
+                res.raise_for_status()
+
+                is_observation = res.json().get("isObservation", False)
+                split_groups = res.json().get("splitGroups", {})
+
+                subset_group_names = []
+                rest_group_names = []
+
+                for group in split_groups:
```
[question]
Is it ok `res.json().get("splitGroups")` returns dict (`{}`)?
If it is dict, the `group` in `for group in split_groups:` would be keys in dict.

Konboi replied:
Ops, splitGroups is an array. I'll fix default value

shibuiwilliam replied:
ok

shibuiwilliam commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +250,76 @@ def run(self):
+                subset_group_names = []
+                rest_group_names = []
+
+                for group in split_groups:
+                    group_name = group.get("groupName", "")
+                    subset = group.get("subset", [])
+                    rests = group.get("rest", [])
+
+                    if is_observation:
+                        subset, rests = subset + rests, []
+
+                    if is_output_exclusion_rules:
+                        for g in split_groups:
+                            if g.get("groupName", "") != group_name:
```
[question]
Is it correct that you do `for g in split_groups` inside of `for group in split_groups`?

Konboi replied:
I want to get all subsets and rests of other groups. Is there any good way to get it without loops?

shibuiwilliam replied:
If that is the case, then this loop is fine.
Another way is to write in list comprehension, which tends to be faster than usual for loop in Python, but not mandatory.

Maybe something like this (I didn't tested):

```python
rests += [g.get("subset", []) + g.get("rest", []) for g in split_groups if g.get("groupName", "") != group_name]
```

shibuiwilliam replied:
Anyway you don't have to fix here if you feel unnecessary.

Konboi replied:
one line is a little difficult to read (in my case), so I'll keep it. Thanks

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -44,6 +49,31 @@
+)
+@click.option(
+    "--split-by-groups-with-rest",
+    'is_split_by_groups_with_rest',
```
The spec was using --rest instead of adding this. Is there a reason for not going with the spec?

Konboi replied:
yes, originally set rest output **file** path to the `--rest` option. Also, needs to set something value when we use the `—rest` option. e.g) launchable-rest.txt

But in split by group should produce `{OUTPUT_DIR}/rest-${GROUP_NAME}.txt`. In my understanding Set `OUTPUT_DIR` path by `---split-by-groups-output-dir` options

So, if we keep the `—rest` option, the user needs to set a dummy path `—rest hoge` to use the rest option.

draftcode replied:
Got it. Updated the spec with the rationale.

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +250,76 @@ def run(self):
+                    if is_output_exclusion_rules:
+                        for g in split_groups:
+                            if g.get("groupName", "") != group_name:
+                                rests = rests + g.get("subset", []) + g.get("rest", [])
```
This part is not needed. As discussed in the spec, we don't consider this case for now. You can just use the `rests` as is without collecting all other rests.

Konboi replied:
Ooo, I had recognized it incorrectly...

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +250,76 @@ def run(self):
+                        subset_group_names.append(group_name)
 
-                self.write_file(rest, rests)
+                    if len(rests) > 0 and group_name != SPLIT_BY_GROUPS_NO_GROUP_NAME:
```
`rest_group_names` should be the complement to `subset_group_names`. This creates an overlap.

draftcode commented on launchable/commands/split_subset.py:
```diff
@@ -180,18 +250,71 @@ def run(self):
+                        subset_group_names.append(group_name)
 
-                self.write_file(rest, rests)
+                    if group_name != SPLIT_BY_GROUPS_NO_GROUP_NAME:
```
`elif`?

Konboi replied:
When the `--output-exclusion-rules` option enable, subset group and rest group will switch. Then, there is rest data, but if it's not in rest_group_names, it's difficult to create subset_group_names.txt
So I added filter logic https://github.com/launchableinc/cli/pull/502/files/8bd7e154a0063632b2a3cf74640024c442cf89c9#diff-36cadddae14cddbb1f42e166c7cb78bb91deeaf3761052e9438363d14dd2c2cbR257-R260

Konboi replied:
I misunderstood so I fixed

Konboi commented:
@draftcode PTAL 🙇
